docs/howtos/when-to-use-what-in-the-ffi.md

```diff
@@ -0,0 +1,130 @@
+
+# When to use what method of passing data between Rust and Java/Swift
+
+There are a bunch of options here. For the purposes of our discussion,
+there are two kinds of values you may want to pass over the FFI.
+
+1. Types with identity
+2. Plain ol' data.
+
+## Types with identity
+
+Examples of this are things like database connections, the FirefoxAccounts
+struct, etc. These types are complex, implemented in rust, and it's not
+unreasonable for them to come to Java/Kotlin as a type representing a
+resource (e.g. implementing `Closable`/`AutoClosable`).
+
+You have two choices here:
+
+1. Use a `ConcurrentHandleMap` to store all instances of your object, and
+   pass the handle back and forth as a u64 from Rust / Long from Kotlin.
+
+   This is recommended for most cases, as it's the hardest to mess up.
+   Additionally, for types T such that `&T: Sync + Send`, or that you
+   need to call `&mut self` method, this is the safest choice.
+
+   Additionally, this will ensure panic-safety, as you'll poison your Mutex.
+
+   The [`ffi_support::handle_map` docs](https://docs.rs/ffi-support/*/ffi_support/handle_map/index.html)
+   are good, and under `ConcurrentHandleMap` include an example of how to set
+   this up. You can also look at most of the FFI crates, as they do this (with
+   the exception of `rc_log`, which has unique requirements).
+
+2. Using an opaque pointer. This is generally only recommended for rare cases
+   like the `PlacesInterruptHandle` (or the `LogAdapterState` from `rc_log`,
+   although it will probably eventually use a handle).
+
+   It's good if your synchronization or threading requirements are somewhat
+   complex and handled separately, such that the additional overhead of
+   the `ConcurrentHandleMap` is undesirable. You should probably talk to us
+   before adding another type that works this way, to make sure it's sound.
+
+   The [`ffi_support` docs](https://docs.rs/ffi-support/*/ffi_support/macro.implement_into_ffi_by_pointer.html) discuss how to do this, or
+   take a look at how it's done for `PlacesInterruptHandle`).
+
+## Plain Old Data
+
+This includes both primitive values, strings, arrays, or arbitrarially nested
+structures containing them.
+
+### Primitives
+
+Specifically numeric primitives. these we'll tackle first since they're the
+easiest.
+
+In general, you can just pass or them as you wish. There are a couple of
+exceptions/caveats. All of them are caused by JNA/Android issues (Swift has very
+good support for calling over the FFI), but it's our lowest common denominator.
+
+1. `bool`: Don't use it. JNA doesn't handle it well. Instead, use a numeric type
+    (like `u8`) and represent 0 for false and 1 for true.
```

Comment:
IIUC we also want to convert it back to bool on the Java side of the FFI wrapper?